### PR TITLE
Detailed Protocol Documentation for XTHINs

### PR DESCRIPTION
**Pull request by ptschip**
still a work in progress but much is already done.

### CODE REVIEW

**Comment by ptschip** [comment]
taking this out of WIP

File: doc/bu-xthin-protocol.md [hunk 1]
```diff
@@ -0,0 +1,382 @@
...
+purged from memory or they may have been unannounced). An XBLOCKTX message which contains the transactions is then sent back.
+
+Generally only one round trip is required to complete the block, but on occasion a transaction has to be re-requested initiating a 
+second round trip.
+
+In addition to the above, the following functionalities and configurations will be needed.
+
+- An XTHIN service bit
+
+- If the thinblocks feature is turned off then thinblocks will not be downloaded but requests for thinblocks will still be serviced.
```
Comment (dagurval) [inline doc/bu-xthin-protocol.md]: What do you mean by turn off the feature?

When turned off, must thinblock requests still be served?

File: doc/bu-xthin-protocol.md [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+
+- The coinbase transaction will always be included in the thinblock.
+
+- During startup when the memory pool has few transactions in it, or when a block is very small and has only 1 or 2 transactions a 
+thinblock may end up being larger than the regular block. In that case a regular block will be returned to the requestor instead of a
```
Comment (dagurval) [inline doc/bu-xthin-protocol.md]: > In that case a regular block will be returned

Is this a hard requirement or a suggestion? This is possibly a recommendation?

I notice that BIPS follow this standard:
The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119.

It is very useful to avoid such confusion.

File: doc/bu-xthin-protocol.md [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+
+- If the thinblocks feature is turned off then thinblocks will not be downloaded but requests for thinblocks will still be serviced.
+
+- The coinbase transaction will always be included in the thinblock.
+
+- During startup when the memory pool has few transactions in it, or when a block is very small and has only 1 or 2 transactions a 
+thinblock may end up being larger than the regular block. In that case a regular block will be returned to the requestor instead of a 
+thinblock. This typically happens when a new block is mined just seconds after the previous one.
+
+- Bloom Size Decay algorithm: A useful phenomena occurs as the memory pools grow and get closer in sync; the bloom filter can be allowed
```
Comment (dagurval) [inline doc/bu-xthin-protocol.md]: Is this a requirement, recommendation or a suggestion? IMO it is up to each node to send whatever bloom filter they fancy.

File: doc/bu-xthin-protocol.md [hunk 4]
```diff
@@ -0,0 +1,382 @@
...
+-----------------
+
+defined in protocol.h.  Add the following service:
+
+    /** nServices flags */
+    enum {
+
+        NODE_XTHIN = (1 << 4),
+
+    ... }
```
Comment (dagurval) [inline doc/bu-xthin-protocol.md]: No need to mention where the enum is defined. It should be enough just to say:

"a node signals that it supports serving xthin blocks using sevice bit 

`NODE_XTHIN = (1 << 4)`

"

File: doc/bu-xthin-protocol.md [hunk 5]
```diff
@@ -0,0 +1,382 @@
...
+    under:
+
+    QString formatServicesStr(quint64 mask)
+
+    add the following:
+
+            case NODE_XTHIN:
+                strList.append("XTHIN");
+                break;
+
```
Comment (dagurval) [inline doc/bu-xthin-protocol.md]: GUI specifics for Bitcoin Unlimited is irrelevant

File: doc/bu-xthin-protocol.md [hunk 6]
```diff
@@ -0,0 +1,382 @@
...
+    ... 
+
+    const char *THINBLOCK="thinblock";
+    const char *XTHINBLOCK="xthinblock";
+    const char *XBLOCKTX="xblocktx";
+    const char *GET_XBLOCKTX="get_xblocktx";
+    const char *GET_XTHIN="get_xthin";
+
+};
```
Comment (dagurval) [inline doc/bu-xthin-protocol.md]: I suggest you don't show C++ code here. Just name them by their protocol string. "thinblock", "xthinblock" etc.

An implementor may not be using C++ nor want to use constants.

File: doc/bu-xthin-protocol.md [hunk 7]
```diff
@@ -0,0 +1,382 @@
...
+
+
+
+Order and logic of P2P message requests
+---------------------------------------
+
+- `NODE A:`  receives block announcement from NODE B by an INV message and then creates and issues a GET_XTHIN to NODE B
+
+- `NODE B:`  receives GET_XTHIN and loads the attached bloom filter and creates an XTHINBLOCK which is sent back to NODE A
+         (the xthinblock is checked before sending back to see if a hash collision is within the block, if so then send a THINBLOCK instead)
```
Comment (dagurval) [inline doc/bu-xthin-protocol.md]: Is checking for hash collision required? Is it protocol violation to send a block with a hash collision?

It should be OK to send a full block also, not only THINBLOCK

File: doc/bu-xthin-protocol.md [hunk 8]
```diff
@@ -0,0 +1,382 @@
...
+
+- `NODE A:`  receives an XTHINBLOCK with 64 bit txn hashes
+    - if there is a hash collision with the mempool then issue GETDATA(MSG_THINBLOCK) to NODE B
```
Comment (dagurval) [inline doc/bu-xthin-protocol.md]: again, may jump directly to MSG_BLOCK

File: doc/bu-xthin-protocol.md [hunk 9]
```diff
@@ -0,0 +1,382 @@
...
+
+
+The Importance of the Orphan Cache
+----------------------------------
+
+It was found that the orphan cache can play an important role in the "thinness" of an xthinblock.  The current core configuration for orphan caching
+is broken.  It is too small and evictions happen to frequently to be of any use.  We found two things of major importance in relation to xthinblocks
+and those are:
+
+- do not purge orphans when a node disconnects.  Currently in Bitcoin Core, when a node disconnects then all orphans that were received from it are
```
Comment (dagurval) [inline doc/bu-xthin-protocol.md]: IMO mentioning Core is irrelevant. Also, it may be useful to mention what orphans are.

File: doc/bu-xthin-protocol.md [hunk 10]
```diff
@@ -0,0 +1,382 @@
...
+A simple bloom filter is used as a compact way to make the node we are requesting an Xthinblock from aware of the contents of our own memory pool. Before
+sending the `GET_XTHIN` message request we create a bloom filter by first getting all the transaction hashes in our own memory pool as well as the
+transaction hashes in the Orphan Cache and then add them to our filter one by one.  The process is very fast and takes just a few milliseconds even
+for a large memory pool.  Once the bloom filter is created, it is appended to the `GET_XTHIN` message before being sent.
+
+Once the receving node receives the `GET_XTHIN` message is removes the filter and stores it in memory as part of the CNode data structure.
+
+        CBloomFilter* pThinBlockFilter;
+
+Storing the filter is important because if there is a hash collision the requester then doesn't have to rebuild the bloom filter before requesting
```
Comment (dagurval) [inline doc/bu-xthin-protocol.md]: Is it possible to send a THINBLOCK request without sending a filter?

File: doc/bu-xthin-protocol.md (hunk 1 above)
Comment (ptschip) [inline doc/bu-xthin-protocol.md]: the feature is the service bit...(i'll make that more clear) but even that doesn't maybe make sense...i think, as far as the functionality, i was thinking in terms of being backwardly compatible in the future and reclaiming the service bit but maybe that's not realistic...maybe we should really have either end be able to turn it off completely, meaning if you are not running with bit ON then some other peer can't still request and xthin from you.  I supposes it's each implementation can do their own thing, it doesn't affect anyone else's...but it does make it more clear perhaps...thoughts?

File: doc/bu-xthin-protocol.md (hunk 2 above)
Comment (ptschip) [inline doc/bu-xthin-protocol.md]: thanks...that's more clear.

File: doc/bu-xthin-protocol.md (hunk 3 above)
Comment (ptschip) [inline doc/bu-xthin-protocol.md]: ok, added some clarity there

File: doc/bu-xthin-protocol.md (hunk 4 above)
Comment (ptschip) [inline doc/bu-xthin-protocol.md]: ok

File: doc/bu-xthin-protocol.md (hunk 5 above)
Comment (ptschip) [inline doc/bu-xthin-protocol.md]: why?

Comment (ptschip) [inline doc/bu-xthin-protocol.md]: actually never mind...some implementations may not be using this...

File: doc/bu-xthin-protocol.md (hunk 6 above)
Comment (ptschip) [inline doc/bu-xthin-protocol.md]: ok, done...

File: doc/bu-xthin-protocol.md (hunk 7 above)
Comment (ptschip) [inline doc/bu-xthin-protocol.md]: done

File: doc/bu-xthin-protocol.md (hunk 8 above)
Comment (ptschip) [inline doc/bu-xthin-protocol.md]: done

File: doc/bu-xthin-protocol.md (hunk 9 above)
Comment (ptschip) [inline doc/bu-xthin-protocol.md]: done

File: doc/bu-xthin-protocol.md (hunk 10 above)
Comment (ptschip) [inline doc/bu-xthin-protocol.md]: added explanation

**Comment by sickpig** [comment]
Now that this got merged could we close it issue #70? 

@TomZ are there any remaining issues?

**Comment by zander** [comment]
Sickpig asked me to look through this as its merged and not a draft anymore.
- the topics have double underlines, which show up in the github rendering of the page. Maybe something to remove?
- the message-types GET_XBLOCKTX and others may be useful to write like `code` even in the highlevel overview.
- additionally, the actual on-the-wire text should be used for this. Not the variable name in your code, its a specification of the spec, not of your implementation.
- The description of the struct is very terse and leaves too much room for interpretation. Frankly, I have no clue what this struct is actually describing.

> // Message format
> //  (4) message start
> //  (12) command
> //  (4) size
> //  (4) checksum
> //  (x) data
- New Enumerated message types
  You need context... Really, this just explains 2 integer values and no idea where to use it and why we should use it.
- The listing of the p2p message formats again talks about your implementation and the variables you used. Please don't describe your implementation, describe the protocol.
- Similarly, your quote "The following illustrates the coding required for this:" shows the same problem where anyone wanting to implement it is unable to see what the on-the-wire data format is. I suggest creating a table of all the fields, their types and the amount of bytes for each.

Example for this is that you write;
    std::vector<CTransaction> vMissingTx;   

What you want to do is explain how that is stored. So one varint and then you can link to something like https://bitcoinfactswiki.github.io/Transaction/ for more.
- NetMsgType::THINBLOCK

I would suggest that THINBLOCK is a completely separate spec and you can refer to it from xthin, but since its stand alone the inclusion into xthin just makes it confusing and bigger.
- "receives an XTHINBLOCK with 64 bit txn hashes"
  You need to specify exactly how that hash is created.

Bottom line; this is a description of your work, not a specification. A 3rd party should be able to implement this in his go-lang based bitcoin client without looking at your code.
